lib/src/DatePicker/Calendar.jsx

```diff
@@ -62,26 +68,46 @@ export class Calendar extends Component {
       date, minDate, maxDate, utils, disableFuture, disablePast,
     } = this.props;
 
-    if (this.shouldDisableDate(date)) {
+    date.forEach(day =>
+      this.shouldDisableDate(day) &&
       this.onDateSelect(findClosestEnabledDate({
-        date,
+        day,
         utils,
         minDate,
         maxDate,
         disablePast,
         disableFuture,
         shouldDisableDate: this.shouldDisableDate,
-      }), false);
-    }
+      }), false)
+    );
   }
 
   onDateSelect = (day, isFinish = true) => {
-    const { date, utils } = this.props;
+    const { date, utils, multi, range } = this.props;
 
-    const withHours = utils.setHours(day, utils.getHours(date));
-    const withMinutes = utils.setMinutes(withHours, utils.getMinutes(date));
+    let newDate = day
+    if (date.length > 0) {
+      newDate = utils.setHours(day, utils.getHours(date[0]));
+      newDate = utils.setMinutes(newDate, utils.getMinutes(date[0]));
+    }
+
+    if (multi) {
```

Comment:
Might this work better as a dictionary? I'm not sure if this would work across the date util libraries, but something like this has worked for my implementation elsewhere:
```
selection = {
  ...selection,
  [day]: !selection[day],
};
```

Reply:
You are correct that a dictionary would be more efficient for multi select mode but `map` and a few other functions of arrays allows us to reuse most of the code for all three modes. As far as I see the small efficiency gain of a dictionary would not be worth the added complexity.